lib/jellyfish_web/server_socket.ex

```diff
@@ -124,33 +114,59 @@ defmodule JellyfishWeb.ServerSocket do
     {:stop, :closed, {1003, "operation not allowed"}, state}
   end
 
-  defp handle_subscribe(
-         %SubscribeRequest{
-           id: id,
-           event_type: {topic, event_type}
-         },
+  defp handle_message({:room_state_request, %RoomStateRequest{room_id: room_id}}, state) do
+    case Room.request_state(room_id) do
+      :ok ->
+        {:ok, {:ok, state}}
+
+      {:error, :room_not_found} ->
+        msg =
+          %ServerMessage{content: {:room_not_found, %RoomNotFound{room_id: room_id}}}
+          |> ServerMessage.encode()
+
+        {:ok, {:reply, :ok, {:binary, msg}, state}}
+    end
+  end
+
+  defp handle_message(
+         {:subscribe_request, %SubscribeRequest{id: id, event_type: {event_type, _event}}},
          state
        )
-       when topic in @valid_subscribe_topics do
-    unless MapSet.member?(state.subscriptions, topic) do
-      :ok = Phoenix.PubSub.subscribe(Jellyfish.PubSub, Atom.to_string(topic))
-    end
+       when event_type in [:server_notification, :metrics] do
+    state = ensure_subscribed(event_type, state)
+
+    reply =
+      %ServerMessage{content: {:subscribe_response, %SubscribeResponse{id: id}}}
+      |> ServerMessage.encode()
+
+    {:ok, {:reply, :ok, {:binary, reply}, state}}
+  end
 
-    state = update_in(state.subscriptions, &MapSet.put(&1, topic))
+  defp handle_message({:get_room_ids, %GetRoomIds{}}, state) do
+    RoomService.request_all_room_ids()
+    {:ok, {:ok, state}}
+  end
 
-    content =
-      case event_type do
-        %SubscribeRequest.ServerNotification{room_id: {_variant, option}} ->
-          get_room_state(option)
+  defp handle_message(message, _state) do
+    {:error, message}
+  end
 
-        %SubscribeRequest.Metrics{} ->
-          nil
-      end
+  @impl true
+  def handle_info({:room_state, room_state}, state) do
+    room_state = to_room_state_message(room_state)
+    msg = %ServerMessage{content: {:room_state, room_state}} |> ServerMessage.encode()
 
-    {:ok, %SubscribeResponse{id: id, content: content}, state}
+    {:reply, :ok, {:binary, msg}, state}
```

Comment:
I don't see this kind of tuple in `Phoenix.Socket.Transport`'s `handle_info` spec on our current Phoenix version. I guess this works because of backwards compatibility, but let's change it to the version described by the spec:

```suggestion
    {:push, {:binary, msg}, state}
```

Reply:
nice find!